c10/core/TensorImpl.cpp

```diff
@@ -81,11 +83,12 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
     // UndefinedTensorImpl is a singleton, so we skip logging it
     C10_LOG_API_USAGE_ONCE("tensor.create");
   }
-  // After we removed Autograd keys from globally enabled set, every Tensor must be created with
-  // a backend DispatchKey and an AutogradBackend key.
-  // We automatically add the corresponding autograd key to key_set_ so that backends can stay
-  // in the old way of only registering with backend key like DispatchKey::CPU.
-  if (c10::InferenceMode::is_enabled()) {
+
+  bool inference_mode = c10::InferenceMode::is_enabled();
+
+  // Inference tensor doesn't have autograd related keys.
+  // View tensor preserves key_set from their base.
+  if (is_view || inference_mode) {
```

Comment:
The key thing here, which I didn't get from the comments above, is that if we are constructing a view of an inference tensor, we MUST NOT put autograd key set on the key set because that would result in a non inference tensor view.

In general, this code is all a bit hard to understand because the meaning of `key_set` is overloaded; for non-views, it is guaranteed not to have any autograd keys and we are responsible for adding them; for views, however, we are given a fully fledged key set and are expected to preserve it as much as possible. It seems like it would be easier to understand if we had two separate constructors, one for non-views, and one for views. (This would also nicely solve the `bool is_view` problem that I've been complaining about elsewhere on the PR).